src/coprocessor/read_pool_impl.rs

```diff
@@ -0,0 +1,144 @@
+}
+
+pub struct ReadPoolImpl;
+
+impl std::fmt::Debug for ReadPoolImpl {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt.debug_struct("coprocessor::ReadPoolImpl").finish()
+    }
+}
+
+impl ReadPoolImpl {
+    pub fn build_read_pool(
+        config: &readpool::Config,
+        pd_sender: FutureScheduler<PdTask>,
+        name_prefix: &str,
+    ) -> ReadPool {
+        let pd_sender2 = pd_sender.clone();
+
+        Builder::from_config(config)
+            .name_prefix(name_prefix)
+            .on_tick(move || ReadPoolImpl::thread_local_flush(&pd_sender))
+            .before_stop(move || ReadPoolImpl::thread_local_flush(&pd_sender2))
+            .build()
+    }
+
+    #[inline]
+    fn thread_local_flush(pd_sender: &FutureScheduler<PdTask>) {
+        // Flush Prometheus metrics
+        LOCAL_COPR_REQ_HISTOGRAM_VEC.with(|m| m.borrow_mut().flush());
+        LOCAL_OUTDATED_REQ_WAIT_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_HANDLE_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_WAIT_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_ERROR.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_SCAN_KEYS.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_ROCKSDB_PERF_COUNTER.with(|m| m.borrow_mut().flush());
+
+        // Report PD metrics
+        LOCAL_COP_FLOW_STATS.with(|local_cop_flow_stats| {
+            if local_cop_flow_stats.borrow().is_empty() {
+                // Stats to report to PD is empty, ignore.
+                return;
+            }
+
+            let read_stats = local_cop_flow_stats.replace(HashMap::default());
+            let result = pd_sender.schedule(PdTask::ReadStats { read_stats });
+            if let Err(e) = result {
+                error!("Failed to send cop pool read flow statistics"; "err" => ?e);
+            }
+        });
+
+        LOCAL_COPR_SCAN_DETAILS.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_GET_OR_SCAN_COUNT.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_EXECUTOR_COUNT.with(|m| m.borrow_mut().flush());
+    }
+
+    pub fn collect(region_id: u64, type_str: &str, metrics: ExecutorMetrics) {
+        let stats = &metrics.cf_stats;
+        // cf statistics group by type
+        for (cf, details) in stats.details() {
+            for (tag, count) in details {
+                LOCAL_COPR_SCAN_DETAILS.with(|m| {
+                    m.borrow_mut()
+                        .with_label_values(&[type_str, cf, tag])
+                        .inc_by(count as i64);
+                });
+            }
+        }
+        // flow statistics group by region
+        ReadPoolImpl::thread_local_collect_read_flow(region_id, stats);
+
+        // scan count
+        //metrics.scan_counter.consume(&mut LOCAL_COPR_GET_OR_SCAN_COUNT);
+        let metrics2 = metrics.clone();
+        LOCAL_COPR_GET_OR_SCAN_COUNT.with(|m| metrics.scan_counter.consume(&mut m.borrow_mut()));
```

Comment:
I prefer to remove this clone and change the interface if `xxx.consume()` is not flexible enough now. Maybe we can simply write `pub range` and `pub point` of `ScanCounter`.